telepresence/startup.py

```diff
@@ -128,6 +118,22 @@ def set_kube_command(runner: Runner, args) -> None:
             )
     context = args.context
 
+    # Add context to prelim_command as it should always be used from now on.
+    # If the user has selected a different context, all version checks need
+    # to use the context the user has selected.
+
+    prelim_command = "{} --context {}".format(prelim_command, context)
```

Comment:
Why does this work? Haven't tried it, but I would expect something like 

No such file or directory: 'kubectl --context default'

Reply:
The preceding command checks and set's the `prelim_command`.
https://github.com/telepresenceio/telepresence/blob/db0ffc1aaf89e26b24c2edc86a83e3f4965c5aac/telepresence/startup.py#L96

If the command does not exist, then it would fail at that check. After the `prelim_command` had been verified, I then I append the --context to the command for all subsequent operations, which is totally valid.

So, with no other flag options specified, it would display the help (two following examples)
## Example 1
```shell
$  kubectl --context docker-for-desktop
kubectl controls the Kubernetes cluster manager.

 Find more information at: https://kubernetes.io/docs/reference/kubectl/overview/

Basic Commands (Beginner):
  create         Create a resource from a file or from stdin.
  expose         Take a replication controller, service, deployment or pod and expose it as a new Kubernetes Service
  run            Run a particular image on the cluster
  set            Set specific features on objects
...
```

## Example 2
```shell
$ kubectl --context default
kubectl controls the Kubernetes cluster manager.

 Find more information at: https://kubernetes.io/docs/reference/kubectl/overview/

Basic Commands (Beginner):
  create         Create a resource from a file or from stdin.
  expose         Take a replication controller, service, deployment or pod and expose it as a new Kubernetes Service
  run            Run a particular image on the cluster
  set            Set specific features on objects
...
```

However, I don't see where in the startup it's running `prelim_command` as a stand alone without any other flags.

# This fix
What this fixes is the following:
Currently, by calling the `api-versions` command without using the context flag the user passes in, and their default context is not working then this is what happens:
## Broken
```shell
$ kubectl api-versions   
Error: unable to initialize DigitalOcean api client: access token is required. (hint: run 'doctl auth init')
error: couldn't get available api versions from server: Get https://{...}.k8s.ondigitalocean.com/api?timeout=32s: getting credentials: exec: exit status 1
```

By setting the context after checking what command to use and that it's valid, all subsequent command will use it and properly succeed.

## Fixed
```shell
$ kubectl --context docker-for-desktop api-versions
admissionregistration.k8s.io/v1beta1
apiextensions.k8s.io/v1beta1
apiregistration.k8s.io/v1
apiregistration.k8s.io/v1beta1
apps/v1
apps/v1beta1
apps/v1beta2
authentication.k8s.io/v1
authentication.k8s.io/v1beta1
authorization.k8s.io/v1
authorization.k8s.io/v1beta1
autoscaling/v1
autoscaling/v2beta1
autoscaling/v2beta2
batch/v1
batch/v1beta1
certificates.k8s.io/v1beta1
compose.docker.com/v1alpha3
compose.docker.com/v1beta1
compose.docker.com/v1beta2
coordination.k8s.io/v1
coordination.k8s.io/v1beta1
events.k8s.io/v1beta1
extensions/v1beta1
networking.k8s.io/v1
networking.k8s.io/v1beta1
node.k8s.io/v1beta1
policy/v1beta1
rbac.authorization.k8s.io/v1
rbac.authorization.k8s.io/v1beta1
scheduling.k8s.io/v1
scheduling.k8s.io/v1beta1
storage.k8s.io/v1
storage.k8s.io/v1beta1
v1
```

Reply:
At a high level your fix makes sense. Thank you for that! I'm wondering about a little detail:

```shell
$ kubectl get pods
No resources found.

$ "kubectl --context default" get pods
bash: kubectl --context default: command not found...
```

That's the shell but I expect Python to behave the same way. You seem to be passing `kubectl --context whatever` as a single string rather than a list of strings. I don't see how that could work. I may very well be missing something, in which case I apologize!

Reply:
I see what you mean. I'll update the PR in a couple of minutes with an appropriate solution.

Reply:
I've updated the `prelim_command` to pass a list of strings. Alternatively, if you prefer, I can manually add the context to each operation.

```python
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from subprocess import Popen
>>> prelim_command = 'kubectl'
>>> context = 'docker-for-desktop'
>>> prelim_command = [prelim_command, '--context', context]
>>> Popen([*prelim_command, "api-versions"])
<subprocess.Popen object at 0x1058a9810>
>>> admissionregistration.k8s.io/v1beta1
apiextensions.k8s.io/v1beta1
apiregistration.k8s.io/v1
apiregistration.k8s.io/v1beta1
apps/v1
apps/v1beta1
apps/v1beta2
authentication.k8s.io/v1
authentication.k8s.io/v1beta1
authorization.k8s.io/v1
authorization.k8s.io/v1beta1
autoscaling/v1
autoscaling/v2beta1
autoscaling/v2beta2
batch/v1
batch/v1beta1
certificates.k8s.io/v1beta1
compose.docker.com/v1alpha3
compose.docker.com/v1beta1
compose.docker.com/v1beta2
coordination.k8s.io/v1
coordination.k8s.io/v1beta1
events.k8s.io/v1beta1
extensions/v1beta1
networking.k8s.io/v1
networking.k8s.io/v1beta1
node.k8s.io/v1beta1
policy/v1beta1
rbac.authorization.k8s.io/v1
rbac.authorization.k8s.io/v1beta1
scheduling.k8s.io/v1
scheduling.k8s.io/v1beta1
storage.k8s.io/v1
storage.k8s.io/v1beta1
v1
```

Reply:
@ark3 I've updated and squashed the commits, good catch on that. I tested the change and seems to work. All I'm doing is setting a new list with the arguments and then unpack the arguments within the command to be called.

Alternatively, I can manually add the context to all the commands, but I think as is looks good and when you read the code you know what is happening.

## Example
```python
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from subprocess import Popen
>>> prelim_command = 'kubectl'
>>> context = 'docker-for-desktop'
>>> prelim_command = [prelim_command, '--context', context]
>>> Popen([*prelim_command, "api-versions"])
<subprocess.Popen object at 0x101d11790>
>>> 
>>> admissionregistration.k8s.io/v1beta1
apiextensions.k8s.io/v1beta1
apiregistration.k8s.io/v1
apiregistration.k8s.io/v1beta1
apps/v1
apps/v1beta1
apps/v1beta2
authentication.k8s.io/v1
authentication.k8s.io/v1beta1
authorization.k8s.io/v1
authorization.k8s.io/v1beta1
autoscaling/v1
autoscaling/v2beta1
autoscaling/v2beta2
batch/v1
batch/v1beta1
certificates.k8s.io/v1beta1
compose.docker.com/v1alpha3
compose.docker.com/v1beta1
compose.docker.com/v1beta2
coordination.k8s.io/v1
coordination.k8s.io/v1beta1
events.k8s.io/v1beta1
extensions/v1beta1
networking.k8s.io/v1
networking.k8s.io/v1beta1
node.k8s.io/v1beta1
policy/v1beta1
rbac.authorization.k8s.io/v1
rbac.authorization.k8s.io/v1beta1
scheduling.k8s.io/v1
scheduling.k8s.io/v1beta1
storage.k8s.io/v1
storage.k8s.io/v1beta1
v1
```